Allow hand-parsed minutes to override regex parser

mrichards42 opened the pull request:
There are a few discrepancies between shapenotecds recordings and the minutes.  Some are from parsing (e.g. missing the memorial lesson), some are from secretaries mis-recording a lead or missing it entirely, and others are likely from transcription errors (e.g. '49b' recorded as '496').

This commit adds hand-parsed minutes for shapenotecds recordings.  Where there was a disagreement between the minutes and recordings, I edited the parsed minutes to reflect the recordings only when there was no ambiguity (e.g. songs recorded as top instead of bottom, or evidence from pictures and recordings that the song or leader was recorded incorrectly).

mrichards42 commented:
I'm not sure what the right way to handle this is.  There are 2 categories of minutes corrections I can think of:

1. Correcting things the parser can't handle, or other difficult to parse constructions.
2. Actual mistakes (incorrectly transcribed numbers, leaders that are missing entirely) that would require a change to the underlying data.

Obviously we should be fixing the first category in the parser eventually, so hand-parsing these minutes isn't a problem.  The second category is definitely trickier.  Some of the included hand-parsed minutes differ from the underlying data based what I could glean from recordings, so while we can know with a pretty high degree of confidence that they are correct and that the minutes themselves are wrong, it still involves changing the data . . . and of course I wasn't considering all this when I initially worked on hand parsing, so both types of changes are included in the PR.

jpkarlsberg commented:
I think we should approach any changes that affect the underlying minutes data in a manner similar to how we presently handle name corrections. Name corrections don't alter underlying data but go into a `leader alias` table. As the minutes are parsed, the raw leader names are mapped through this table to get to a `leader_id`. I think it's important that we not change the underlying data because it's of historical significance, might be useful in unmodified form for other purposes, and might after all be correct (though in the cases at hand, I agree there's a high likelihood the underlying data is wrong).

I think we should strive for similar to this method of handling name corrections when it comes to other kinds errors in the minutes. Accommodating such errors is more complicated than name corrections because there are more kinds of errors, including the ones @mrichards42 mentions (the parsing script not picking up certain songs and songs where the number is wrong). 

We should think about how best to construct a new table for corrections to page numbers (and perhaps other sorts of changes) that can function like the name corrections does. Or is there a better way to handle this that preserves the underlying data?

@marktgodfrey: do you agree that hand-parsing isn't a problem because the parser will eventually be improved?

marktgodfrey commented:
I think the parser will always have problems, and should really be just thought of as a tool to get us to an intermediate format that encodes the raw minutes text into a form that is easily convertible/injectable to SQL (or whatever).  I guess Mike's "hand-parsed" format is a first shot at this format, but it should probably be expanded, right?  @jpkarlsberg I know you had a thread with Will Fitzgerald about a nested sort of format (like a pseudo-JSON if I recall) that could encapsulate all the data and structure of the minutes and could also carry along the raw text and correction info along with it for archival purposes.  This would work well with Mike's interface for hand-correcting the parsed minutes he showed us last year, and I think jibe with Chris Thorman's new work getting all the minutes processing out of spreadsheets and into a database.
I imagine it flowing like this: raw minutes texts from the secretaries goes through parser(s) and/or manual-editing interface to the encoded format that gets stored in a repository.  The SQLite database that drives the apps (or any database anyone wants to make ever) would just need to parse and inject the data from these encoded minutes "files" in a canonical way.
If this sounds OK, how far are we from a sort of "encoded format"?  Mike's version gets the job done but I feel like we could expand it a little and get some high-value archival data passed to it.  Definitely don't want to have a year's worth of meetings about it, think we could get 90% of the value soon with little effort, just need to figure it out and get it over with!
What do y'all think?  Sorry for the stream of consciousness...

mrichards42 commented:
Agreed that the parser will always be a work in progress, and that we should be using a single file/json object/whatever per singing for this intermediary format.  And I'm on board with Mark's suggestion that we go minutes -> parser -> some intermediary format in a repository, and then anything that wants to consume data reads from the repo and dumps into, e.g. a database.  Once we have historical data involved it might be impractical to ship the entire db with the app, so we'll probably want a web api for querying the repository.

As for the intermediate format, I would prefer something that is easy to edit by hand rather than requiring an interface to use, just so we could get started using it quickly.  If we get more people to help, we'll probably want some kind of interface so less technical people can contribute without worrying about syntax.

Fwiw, I wouldn't consider csv a first shot at the format, it was just an easy way to dump the data and edit it, and easy to make the parser understand.

I'd rather something like json or yaml (since you can't really store the full minutes text in json in a human-readable format) . . .

Something along these lines:

```yaml
name: Keystone Sacred Harp Convention
location: Broad Street Ministry, Philadelphia, Pennsylvania
date: January 24-25, 2009
meta:
  audio_url: http://phillysacredharp.org/keystone-recordings#2009
  year: 2009
minutes: >
  Saturday, January 24

  The 11th annual Keystone Sacred Harp Convention was convened by Ina Shea and Elizabeth
  Stokes leading [82t]. Lynne Hoyt offered the opening prayer.

  [...]

leads:
- {song: '82t', leaders: [Ina Shea, Elizabeth Stokes]}
- {song: '72b', leaders: [Anna Mays]}
- {song: '394', leaders: [Ruth Wampler]}
- {song: '112', leaders: [G.C. Waldrep]}
- {song: '77b', leaders: [Cindy Wasson]}
[...]
```

And then if you need to make a correction:
```yaml
leads:
# In the actual minutes text, but should not have been there
- {song: '72b', leaders: [Anna Mays], deleted: true, corrected_by: john.doe@example.com}
# In the actual minutes text, but the wrong number
- {song: '39b', leaders: [Ruth Wampler], original: {song: '349'}, corrected_by: jane.doe@example.com}
# Missing from the minutes text
- {song: '377', leaders: [Mike Richards], added: true, corrected_by: jane.doe@example.com}
```

It would be pretty slick to omit the corrected_by field and just store the minutes as their own git repo, but

marktgodfrey commented:
I really like the look of that example yaml  @mrichards42 !

I'm communicating with Chris Thorman to try to coordinate all efforts here.  He's recently done some work building a "Minuted Singing Event" "document" object that seems very related to the format we're discussing here.  I should be meeting with him soon, hopefully will have more clarity on how to best proceed so that we're all moving in the same direction.

mrichards42 commented:
Any news from the meeting with Chris?

(moving the discussion back over here instead of in #11)
